### ISSUE TITLE
Segfault in do_stop_gossip due to double stop in commitlog_test:TestCommitLog.stop_failure_policy_test

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylla#8911]
Seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/3677/artifact/logs-gating.release.2/1624407595877_commitlog_test.TestCommitLog.stop_failure_policy_test/node1.log
but it seems applicable to master as well
```
Scylla version 4.6.dev-0.20210622.4ab4f63efe with build-id 626f1aadc40563f6bab52bf780b258ab27c06cba starting ...
...
ERROR 2021-06-23 00:19:50,255 [shard 1] storage_service - Shutting down communications due to I/O errors until operator intervention: Commitlog error: std::filesystem::__cxx11::filesystem_error (error system:13, filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426080.log])
WARN  2021-06-23 00:19:50,255 [shard 1] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426080.log])
WARN  2021-06-23 00:19:50,355 [shard 1] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426081.log])
INFO  2021-06-23 00:19:50,396 [shard 0] storage_service - Stop transport: starts
INFO  2021-06-23 00:19:50,396 [shard 0] storage_service - Shutting down native transport
INFO  2021-06-23 00:19:50,401 [shard 0] cql_server_controller - CQL server stopped
INFO  2021-06-23 00:19:50,401 [shard 0] storage_service - Shutting down native transport was successful
INFO  2021-06-23 00:19:50,401 [shard 0] storage_service - Shutting down rpc server
INFO  2021-06-23 00:19:50,401 [shard 0] thrift_controller - Thrift server stopped
INFO  2021-06-23 00:19:50,401 [shard 0] storage_service - Shutting down rpc server was successful
INFO  2021-06-23 00:19:50,401 [shard 0] storage_service - Stop transport: shutdown rpc and cql server done
INFO  2021-06-23 00:19:50,401 [shard 0] gossip - My status = NORMAL
INFO  2021-06-23 00:19:50,401 [shard 0] gossip - Announcing shutdown
INFO  2021-06-23 00:19:50,402 [shard 0] storage_service - Node 127.0.93.1 state jump to normal
WARN  2021-06-23 00:19:50,456 [shard 1] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426082.log])
WARN  2021-06-23 00:19:50,556 [shard 1] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426083.log])
WARN  2021-06-23 00:19:50,755 [shard 1] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426084.log])
WARN  2021-06-23 00:19:51,081 [shard 1] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426085.log])
INFO  2021-06-23 00:19:51,095 [shard 1] compaction - [Compact system.peers b9e1ac70-d3b8-11eb-b78e-6d971ecda094] Compacting [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/data/system/peers-37f71aca7dc2383ba70672528af04d4f/md-5-big-Data.db:level=0:origin=memtable, /jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/data/system/peers-37f71aca7dc2383ba70672528af04d4f/md-3-big-Data.db:level=0:origin=memtable, ]
WARN  2021-06-23 00:19:52,342 [shard 1] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426086.log])
INFO  2021-06-23 00:19:52,344 [shard 1] compaction - [Compact system.peers b9e1ac70-d3b8-11eb-b78e-6d971ecda094] Compacted 2 sstables to []. 11kB to 0 bytes (~0% of original) in 0ms = 0 bytes/s. ~256 total partitions merged to 0.
WARN  2021-06-23 00:19:52,345 [shard 0] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-1089944095.log])
WARN  2021-06-23 00:19:52,443 [shard 1] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426087.log])
WARN  2021-06-23 00:19:52,445 [shard 0] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-1089944096.log])
WARN  2021-06-23 00:19:52,543 [shard 1] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426088.log])
WARN  2021-06-23 00:19:52,546 [shard 0] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-1089944097.log])
WARN  2021-06-23 00:19:52,644 [shard 1] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426089.log])
WARN  2021-06-23 00:19:52,646 [shard 0] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-1089944098.log])
WARN  2021-06-23 00:19:52,744 [shard 1] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426090.log])
WARN  2021-06-23 00:19:52,747 [shard 0] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 13: filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-1089944099.log])
INFO  2021-06-23 00:19:52,864 [shard 0] compaction_manager - Asked to stop
INFO  2021-06-23 00:19:52,864 [shard 0] compaction_manager - Stopping 0 ongoing compactions due to shutdown
WARN  2021-06-23 00:19:52,864 [shard 0] gossip - failure_detector_loop: Got error in the loop, live_nodes=[]: seastar::sleep_aborted (Sleep is aborted)
INFO  2021-06-23 00:19:52,864 [shard 0] gossip - failure_detector_loop: Finished main loop
INFO  2021-06-23 00:19:52,864 [shard 0] init - Signal received; shutting down
INFO  2021-06-23 00:19:52,864 [shard 0] init - Shutting down redis service
INFO  2021-06-23 00:19:52,864 [shard 0] init - Shutting down redis service was successful
INFO  2021-06-23 00:19:52,864 [shard 0] init - Shutting down view builder
INFO  2021-06-23 00:19:52,864 [shard 0] view - Stopping view builder
INFO  2021-06-23 00:19:52,864 [shard 0] compaction_manager - Stopped
INFO  2021-06-23 00:19:52,864 [shard 1] compaction_manager - Asked to stop
INFO  2021-06-23 00:19:52,864 [shard 1] compaction_manager - Stopping 0 ongoing compactions due to shutdown
INFO  2021-06-23 00:19:52,865 [shard 1] view - Stopping view builder
INFO  2021-06-23 00:19:52,865 [shard 1] compaction_manager - Stopped
INFO  2021-06-23 00:19:52,865 [shard 0] init - Shutting down view builder was successful
INFO  2021-06-23 00:19:52,865 [shard 0] init - Shutting down local storage
INFO  2021-06-23 00:19:52,865 [shard 0] storage_service - Stop transport: starts
INFO  2021-06-23 00:19:52,865 [shard 0] storage_service - Shutting down native transport
INFO  2021-06-23 00:19:52,865 [shard 0] storage_service - Shutting down native transport was successful
INFO  2021-06-23 00:19:52,865 [shard 0] storage_service - Shutting down rpc server
INFO  2021-06-23 00:19:52,865 [shard 0] storage_service - Shutting down rpc server was successful
INFO  2021-06-23 00:19:52,865 [shard 0] storage_service - Stop transport: shutdown rpc and cql server done
INFO  2021-06-23 00:19:52,865 [shard 0] gossip - My status = shutdown
INFO  2021-06-23 00:19:52,865 [shard 0] gossip - Announcing shutdown
INFO  2021-06-23 00:19:52,865 [shard 0] storage_service - Node 127.0.93.1 state jump to normal
INFO  2021-06-23 00:19:53,095 [shard 0] gossip - Disable and wait for gossip loop started
INFO  2021-06-23 00:19:53,095 [shard 0] gossip - Disable and wait for gossip loop finished
INFO  2021-06-23 00:19:53,096 [shard 0] gossip - Gossip is now stopped

                                        ^^^^^^

INFO  2021-06-23 00:19:53,096 [shard 0] storage_service - Stop transport: stop_gossiping done
INFO  2021-06-23 00:19:53,096 [shard 0] messaging_service - Stopping nontls server
INFO  2021-06-23 00:19:53,096 [shard 0] messaging_service - Stopping tls server
INFO  2021-06-23 00:19:53,096 [shard 0] messaging_service - Stopping tls server - Done
INFO  2021-06-23 00:19:53,096 [shard 1] messaging_service - Stopping nontls server
INFO  2021-06-23 00:19:53,096 [shard 1] messaging_service - Stopping tls server
INFO  2021-06-23 00:19:53,096 [shard 1] messaging_service - Stopping tls server - Done
INFO  2021-06-23 00:19:53,096 [shard 0] messaging_service - Stopping nontls server - Done
INFO  2021-06-23 00:19:53,096 [shard 1] messaging_service - Stopping nontls server - Done
INFO  2021-06-23 00:19:53,096 [shard 0] storage_service - messaging_service stopped
INFO  2021-06-23 00:19:53,096 [shard 0] storage_service - Stop transport: shutdown messaging_service done
INFO  2021-06-23 00:19:53,096 [shard 0] storage_service - stream_manager stopped
INFO  2021-06-23 00:19:53,096 [shard 0] storage_service - Stop transport: shutdown stream_manager done
INFO  2021-06-23 00:19:53,096 [shard 0] storage_service - Stop transport: done

==> why is gossip trying to stop again?
we should have seen `"gossip is already stopped"` from https://github.com/scylladb/scylla/blob/373fa3fa07d366ea5a8632f0ddffae6f226f64c3/gms/gossiper.cc#L2151

INFO  2021-06-23 00:19:54,915 [shard 0] gossip - Disable and wait for gossip loop started
INFO  2021-06-23 00:19:54,915 [shard 0] gossip - Disable and wait for gossip loop finished
Segmentation fault on shard 0.
Backtrace:
  0x405e608
  0x4090256
  0x7ff753d0b1df
  0x290ab60
  0x42dab23
  0x42daaaf
  0x293fa3f
  0x42edca1
```

Core dump: https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/3677/artifact/logs-gating.release.2/1624407595877_commitlog_test.TestCommitLog.stop_failure_policy_test/node1-scylla.11438.1624407594.core.gz
reloc package: http://downloads.scylladb.com/unstable/scylla/master/relocatable/2021-06-22T22:42:30Z/scylla-package.tar.gz

Decoded:
```
(gdb) bt
#0  0x000000000290ab61 in seastar::internal::future_base::detach_promise (this=0x60000523fce8) at ./seastar/include/seastar/core/future.hh:1170
#1  seastar::internal::future_base::schedule (this=0x60000523fce8, tws=0x6000005ec4c0, state=0x6000005ec4d0) at ./seastar/include/seastar/core/future.hh:1175
#2  seastar::future<void>::schedule (this=0x60000523fce8, tws=0x6000005ec4c0) at ./seastar/include/seastar/core/future.hh:1372
#3  seastar::future<void>::schedule<seastar::internal::promise_base_with_type<void>, seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false>, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false> >(seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, auto:1&&)::{lambda()#1}, false>&, seastar::future_state<seastar::internal::monostate>&&)#1}>(seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&, seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false>&&, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false> >(seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, auto:1&&)::{lambda()#1}, false>&, seastar::future_state<seastar::internal::monostate>&&)#1}&&) (this=0x60000523fce8, pr=..., func=..., wrapper=...) at ./seastar/include/seastar/core/future.hh:1391
#4  seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false> >(seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false>&&) (this=0x60000523fce8, func=...) at ./seastar/include/seastar/core/future.hh:1666
#5  seastar::future<void>::then_wrapped_common<false, seastar::future<void>, seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false> >(seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false>&&) (this=0x60000523fce8, func=...) at ./seastar/include/seastar/core/future.hh:1691
#6  seastar::future<void>::then_wrapped_maybe_erase<false, seastar::future<void>, seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false> >(seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false>&&) (this=0x60000523fce8, func=...) at ./seastar/include/seastar/core/future.hh:1644
#7  seastar::future<void>::then_wrapped<seastar::future<void>::finally_body<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}, false>, seastar::future<void> >(seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&) & (this=0x60000523fce8, func=...) at ./seastar/include/seastar/core/future.hh:1628
#8  seastar::future<void>::finally<seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&)::{lambda()#1}>(seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&) (this=0x60000523fce8, func=...) at ./seastar/include/seastar/core/future.hh:1745
#9  seastar::smp::submit_to<seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>(unsigned int, seastar::smp_submit_to_options, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&&) (t=<optimized out>, options=..., func=...) at ./seastar/include/seastar/core/smp.hh:348
#10 seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const (this=<optimized out>, c=<optimized out>) at ./seastar/include/seastar/core/sharded.hh:709
#11 std::__invoke_impl<seastar::future<void>, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}&, unsigned int>(std::__invoke_other, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}&, unsigned int&&) (__f=..., __args=<optimized out>) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:60
#12 std::__invoke_r<seastar::future<void>, seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}&, unsigned int>(seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}&, unsigned int&&) (__fn=..., __args=<optimized out>) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:113
#13 std::_Function_handler<seastar::future<void> (unsigned int), seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>)::{lambda(unsigned int)#1}>::_M_invoke(std::_Any_data const&, unsigned int&&) (__functor=..., __args=<optimized out>) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/std_function.h:291
#14 0x00000000042dab24 in std::function<seastar::future<void> (unsigned int)>::operator()(unsigned int) const (this=0x60000523fe70, __args=0) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/std_function.h:622
#15 seastar::futurize<seastar::future<void> >::invoke<std::function<seastar::future<void> (unsigned int)>, unsigned int>(std::function<seastar::future<void> (unsigned int)>&&, unsigned int&&) (func=..., args=<optimized out>) at ./seastar/include/seastar/core/future.hh:2135
#16 seastar::futurize_invoke<std::function<seastar::future<void> (unsigned int)>, unsigned int>(std::function<seastar::future<void> (unsigned int)>&&, unsigned int&&) (func=..., args=<optimized out>) at ./seastar/include/seastar/core/future.hh:2166
#17 seastar::parallel_for_each<boost::range_detail::integer_iterator<unsigned int>, std::function<seastar::future<void> (unsigned int)> >(boost::range_detail::integer_iterator<unsigned int>, boost::range_detail::integer_iterator<unsigned int>, std::function<seastar::future<void> (unsigned int)>&&) (begin=..., begin@entry=..., end=..., func=...) at ./seastar/include/seastar/core/loop.hh:554
#18 0x00000000042daab0 in seastar::internal::parallel_for_each_impl<boost::integer_range<unsigned int>, std::function<seastar::future<void> (unsigned int)> >(boost::integer_range<unsigned int>&&, std::function<seastar::future<void> (unsigned int)>&&) (range=..., func=...) at ./seastar/include/seastar/core/loop.hh:598
#19 seastar::futurize<seastar::future<void> >::invoke<seastar::future<void> (*&)(boost::integer_range<unsigned int>&&, std::function<seastar::future<void> (unsigned int)>&&), boost::integer_range<unsigned int>, std::function<seastar::future<void> (unsigned int)> >(seastar::future<void> (*&)(boost::integer_range<unsigned int>&&, std::function<seastar::future<void> (unsigned int)>&&), boost::integer_range<unsigned int>&&, std::function<seastar::future<void> (unsigned int)>&&) (func=<optimized out>, args=..., args=...) at ./seastar/include/seastar/core/future.hh:2135
#20 seastar::futurize_invoke<seastar::future<void> (*&)(boost::integer_range<unsigned int>&&, std::function<seastar::future<void> (unsigned int)>&&), boost::integer_range<unsigned int>, std::function<seastar::future<void> (unsigned int)> >(seastar::future<void> (*&)(boost::integer_range<unsigned int>&&, std::function<seastar::future<void> (unsigned int)>&&), boost::integer_range<unsigned int>&&, std::function<seastar::future<void> (unsigned int)>&&) (func=<optimized out>, args=..., args=...) at ./seastar/include/seastar/core/future.hh:2166
#21 seastar::parallel_for_each<boost::integer_range<unsigned int>, std::function<seastar::future<void> (unsigned int)> >(boost::integer_range<unsigned int>&&, std::function<seastar::future<void> (unsigned int)>&&) (range=..., func=...) at ./seastar/include/seastar/core/loop.hh:610
#22 seastar::internal::sharded_parallel_for_each(unsigned int, std::function<seastar::future<void> (unsigned int)>) (nr_shards=6, on_each_shard=...) at ./seastar/src/core/sharded.cc:33
#23 0x000000000293fa40 in seastar::sharded<gms::gossiper>::sharded_parallel_for_each(std::function<seastar::future<void> (unsigned int)>) (this=0x44f6c48 <gms::_the_gossiper>, func=...) at ./seastar/include/seastar/core/sharded.hh:160
#24 seastar::sharded<gms::gossiper>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (gms::gossiper&)>) (this=0x44f6c48 <gms::_the_gossiper>, options=..., func=...) at ./seastar/include/seastar/core/sharded.hh:708
#25 seastar::sharded<gms::gossiper>::invoke_on_all<gms::gossiper::do_stop_gossiping()::$_33::operator()() const::{lambda(auto:1&)#1}>(seastar::smp_submit_to_options, gms::gossiper::do_stop_gossiping()::$_33::operator()() const::{lambda(auto:1&)#1}) (this=0x44f6c48 <gms::_the_gossiper>, options=..., func=...) at ./seastar/include/seastar/core/sharded.hh:727
#26 seastar::sharded<gms::gossiper>::invoke_on_all<gms::gossiper::do_stop_gossiping()::$_33::operator()() const::{lambda(auto:1&)#1}>(gms::gossiper::do_stop_gossiping()::$_33::operator()() const::{lambda(auto:1&)#1}) (this=0x44f6c48 <gms::_the_gossiper>, func=...) at ./seastar/include/seastar/core/sharded.hh:249
#27 gms::gossiper::do_stop_gossiping()::$_33::operator()() const (this=<optimized out>) at gms/gossiper.cc:2199
#28 std::__invoke_impl<void, gms::gossiper::do_stop_gossiping()::$_33>(std::__invoke_other, gms::gossiper::do_stop_gossiping()::$_33&&) (__f=...) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:60
#29 std::__invoke<gms::gossiper::do_stop_gossiping()::$_33>(gms::gossiper::do_stop_gossiping()::$_33&&) (__fn=...) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:95
#30 std::__apply_impl<gms::gossiper::do_stop_gossiping()::$_33, std::tuple<>>(gms::gossiper::do_stop_gossiping()::$_33&&, std::tuple<>&&, std::integer_sequence<unsigned long>) (__t=..., __f=...) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/tuple:1723
#31 std::apply<gms::gossiper::do_stop_gossiping()::$_33, std::tuple<> >(gms::gossiper::do_stop_gossiping()::$_33&&, std::tuple<>&&) (__t=..., __f=...) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/tuple:1734
#32 seastar::futurize<void>::apply<gms::gossiper::do_stop_gossiping()::$_33>(gms::gossiper::do_stop_gossiping()::$_33&&, std::tuple<>&&) (args=..., func=...) at ./seastar/include/seastar/core/future.hh:2099
#33 seastar::async<gms::gossiper::do_stop_gossiping()::$_33>(seastar::thread_attributes, gms::gossiper::do_stop_gossiping()::$_33&&)::{lambda()#1}::operator()() const (this=<optimized out>) at ./seastar/include/seastar/core/thread.hh:258
#34 seastar::noncopyable_function<void ()>::direct_vtable_for<seastar::async<gms::gossiper::do_stop_gossiping()::$_33>(seastar::thread_attributes, gms::gossiper::do_stop_gossiping()::$_33&&)::{lambda()#1}>::call(seastar::noncopyable_function<void ()> const*) (func=0x6000002321a0) at ./seastar/include/seastar/util/noncopyable_function.hh:124
#35 0x00000000042edca2 in seastar::noncopyable_function<void ()>::operator()() const (this=0x6000005ec4d8) at ./seastar/include/seastar/util/noncopyable_function.hh:209
#36 seastar::thread_context::main (this=0x600000232180) at ./seastar/src/core/thread.cc:299
```

**Comment by bhalevy** [comment]
Cc @xemul

**Comment by bhalevy** [comment]
We call `stop_gossiping` both from https://github.com/scylladb/scylla/blob/373fa3fa07d366ea5a8632f0ddffae6f226f64c3/service/storage_service.cc#L1345
and https://github.com/scylladb/scylla/blob/373fa3fa07d366ea5a8632f0ddffae6f226f64c3/main.cc#L1199

But here it looks like both invocations come from `stop_transport` according to the log.

In any case, `stop_gossiping` and `do_stop_gossiping` are supposed to be resilient against multiple invocations.

I guess that the global `distributed<gossiper>` is never `sharded::stop()ed` so we go though in `stop_gossiping`, but then we set `_enabled=false` in https://github.com/scylladb/scylla/blob/373fa3fa07d366ea5a8632f0ddffae6f226f64c3/gms/gossiper.cc#L2188
which we clearly have gone through, again according to the log.

**Comment by bhalevy** [comment]
FWIW, according to the log, it looks like we didn't get to https://github.com/scylladb/scylla/blob/373fa3fa07d366ea5a8632f0ddffae6f226f64c3/main.cc#L1198-L1200

Last deferred shutdown is from https://github.com/scylladb/scylla/blob/373fa3fa07d366ea5a8632f0ddffae6f226f64c3/main.cc#L1418-L1420

**Comment by bhalevy** [comment]
Aha! `storage_service::stop_transport` runs twice, apparently in parallel.
```
INFO  2021-06-23 00:19:50,396 [shard 0] storage_service - Stop transport: starts
INFO  2021-06-23 00:19:50,396 [shard 0] storage_service - Shutting down native transport
INFO  2021-06-23 00:19:50,401 [shard 0] cql_server_controller - CQL server stopped
INFO  2021-06-23 00:19:50,401 [shard 0] storage_service - Shutting down native transport was successful
INFO  2021-06-23 00:19:50,401 [shard 0] storage_service - Shutting down rpc server
INFO  2021-06-23 00:19:50,401 [shard 0] thrift_controller - Thrift server stopped
INFO  2021-06-23 00:19:50,401 [shard 0] storage_service - Shutting down rpc server was successful
INFO  2021-06-23 00:19:50,401 [shard 0] storage_service - Stop transport: shutdown rpc and cql server done
...
INFO  2021-06-23 00:19:52,865 [shard 0] init - Shutting down view builder was successful
INFO  2021-06-23 00:19:52,865 [shard 0] init - Shutting down local storage
INFO  2021-06-23 00:19:52,865 [shard 0] storage_service - Stop transport: starts
```

**Comment by bhalevy** [comment]
I think we go through https://github.com/scylladb/scylla/blob/373fa3fa07d366ea5a8632f0ddffae6f226f64c3/service/storage_service.cc#L3604-L3608

as per
```
ERROR 2021-06-23 00:19:50,255 [shard 1] storage_service - Shutting down communications due to I/O errors until operator intervention: Commitlog error: std::filesystem::__cxx11::filesystem_error (error system:13, filesystem error: open failed: Permission denied [/jenkins/workspace/scylla-master/next@2/scylla/.dtest/dtest-jtk1_gpl/test/node1/commitlogs/CommitLog-2-18014399599426080.log])
```

probably due to https://github.com/scylladb/scylla/pull/8578/commits/f01307d816cf9e0155bd0a835adabc4b977f2219 which seems to be doing what it's supposed to do

**Comment by bhalevy** [comment]
Bottom line is that we need to harden `stop_transport` and `do_stop_gossiping` so they could be invoked in parallel and be idempotent.
